### PR TITLE
feat(dns): Support DNS queryStrategy config per NameServer.

### PR DESCRIPTION
**Pull request by cty123**
There are multiple feature requests already to enable `queryStrategy` configuration at the granularity of DNS name server: #1953 #1940. The PR adds a field `queryStrategy` in NameServer struct to allow user to configure DNS query strategy for every NameServer instead of using a single option for all name servers. 

The configuration before looks likes this,
```json
"dns": {
  "servers": [
    {
      ... NameServer1
    },
    {
      ... NameServer2
    }
  ],
  "queryStrategy": "UseIP6"
}
```

After this PR, users are allowed to specify the `queryStrategy` option for each of the NameServers, for example,
```json
"dns": {
  "servers": [
    {
      ... NameServer1
      +"queryStrategy": "UseIP4"
    },
    {
      ... NameServer2
      +"queryStrategy": "UseIP6"
    }
  ],
  -"queryStrategy": "UseIP6"
}
```

To be compatible with the old configuration,

1. The old `queryStrategy` field applied on the DNS config level still works. The NameServer level `queryStrategy` option can be treated as another filter. This should be pretty straight-forward, for example, in the following case, the queryStrategy would be set to `UseIP6`,
```json
"dns": {
  "servers": [
    {
      ... NameServer1
      "queryStrategy": "UseIP6"
    },
  ],
  "queryStrategy": "UseIP"
}
```

but in the following case, the `queryStrategy` would always result in empty response, due to conflicting DNS option,

```json
"dns": {
  "servers": [
    {
      ... NameServer1
      "queryStrategy": "UseIP6"
    },
  ],
  "queryStrategy": "UseIP4"
}
```

Therefore, users are encouraged to use NameServer level `queryStrategy` field.

2. If `queryStrategy` is not set at NameServer level, the NameService will not override the `queryStrategy` setting. The behavior of the DNS resolver should remain the same as previous.

### CODE REVIEW

**Comment by yuhan6665** [comment]
麻烦看一下 TestLocalNameServerWithIPv6Override 在 windows 这个测试

**Comment by cty123** [comment]
> 麻烦看一下 TestLocalNameServerWithIPv6Override 在 windows 这个测试

可能跟windows下go的DNS lookup实现有关系，macos和linux都能过这个测试。稍微看了下源码，
```go
       if r.preferGoOverWindows() {
		return r.goLookupIP(ctx, network, name)
	}
	// TODO(bradfitz,brainman): use ctx more. See TODO below.

	var family int32 = syscall.AF_UNSPEC
	switch ipVersion(network) {
	case '4':
		family = syscall.AF_INET
	case '6':
		family = syscall.AF_INET6
	}
...
```
有可能跟运行环境有关系，但是目前手上暂时没有windows的机器，所以没法测试。所以暂时移除了在local nameserver使用queryStrategy的功能。

File: app/dns/nameserver_doh.go [hunk 1]
```diff
@@ -408,3 +414,24 @@ func (s *DoHNameServer) QueryIP(ctx context.Context, domain string, clientIP net
 		}
 	}
 }
+
+func (s *DoHNameServer) resolveIpOptionOverride(ipOption dns_feature.IPOption) dns_feature.IPOption {
+	switch s.queryStrategy {
```
[the remark this answers is not in the document]
Comment (cty123) [reply app/dns/nameserver_doh.go]: Created a more general function under `nameserver.go`

**Comment by yuhan6665** [comment]
感谢佬！

**Comment by us254** [comment]
"Is this the correct way to use this method?"



```
{
  "dns": {
    "servers": [
      {
        "name": "NameServer1",
        "address": "1.1.1.1",
        "port": 53,
        "queryStrategy": "UseIP4"
      },
      {
        "name": "NameServer2",
        "address": "8.8.8.8",
        "port": 53,
        "queryStrategy": "UseIP6"
      }
    ]
  },
  "outbounds": [
    {
      "tag": "dns1-outbound",
      "protocol": "dns",
      "settings": {
        "servers": [
          "NameServer1"
        ]
      }
    },
    {
      "tag": "dns2-outbound",
      "protocol": "dns",
      "settings": {
        "servers": [
          "NameServer2"
        ]
      }
    }
  ],
  "routing": {
    "rules": [
      {
        "type": "field",
        "outboundTag": "dns1-outbound",
        "domain": [
          "aa.com",
          "bb.org"
        ]
      },
      {
        "type": "field",
        "outboundTag": "dns2-outbound",
        "domain": [
          "cc.net",
          "dd.org"
        ]
      }
    ]
  }
}
```

**Pull request merged**